charms/kfp-api/src/charm.py

```diff
@@ -286,98 +202,180 @@ def set_pod_spec(self, event):
+            raise CheckFailed("Too many mysql relations", BlockedStatus)
+
+        try:
+            mysql = mysql[0]
+            unit = list(mysql.units)[0]
+            mysql = mysql.data[unit]
+        except Exception as e:
+            self.log.error(
+                f"Encountered the following exception when parsing mysql relation: "
+                f"{str(e)}"
+            )
+            raise CheckFailed(
+                "Unexpected error when parsing mysql relation.  See logs", BlockedStatus
+            )
+
+        expected_attributes = ["database", "host", "root_password", "port"]
+
+        missing_attributes = [attribute for attribute in expected_attributes if attribute not in mysql]
+
+        if len(missing_attributes) == len(expected_attributes):
+            raise CheckFailed("Waiting for mysql relation data", WaitingStatus)
+        elif len(missing_attributes) > 0:
+            self.log.error(
+                f"mysql relation data missing expected attributes '{missing_attributes}'"
+            )
+            raise CheckFailed(
+                "Received incomplete data from mysql relation.  See logs", BlockedStatus
+            )
+        return mysql
+
+    def _get_object_storage(self, interfaces):
+        relation_name = "object-storage"
+        return self.validate_sdi_interface(interfaces, relation_name)
+
+    def _get_viz(self, interfaces):
+        relation_name = "kfp-viz"
+        default_viz_data = {"service-name": "unset", "service-port": "1234"}
+        return self.validate_sdi_interface(interfaces, relation_name, default_return=default_viz_data)
+
+    def validate_sdi_interface(self, interfaces: dict, relation_name: str, default_return=None):
+        """Validates data received from SerializedDataInterface, returning the data if valid
+
+        Optionally can return a default_return value when no relation is established
+
+        Raises:
+            CheckFailed(..., Blocked) when no relation established (unless default_return set)
+            CheckFailed(..., Blocked) if interface is not using SDI
+            CheckFailed(..., Blocked) if data in interface fails schema check
+            CheckFailed(..., Waiting) if we have a relation established but no data passed
+
+        Params:
+            interfaces:
+
+        Returns:
+              (dict) interface data
+        """
+        # If nothing is related to this relation, return a default value or raise an error
+        if relation_name not in interfaces or interfaces[relation_name] is None:
+            if default_return is not None:
+                return default_return
+            else:
+                raise CheckFailed(f"Missing required relation for {relation_name}", BlockedStatus)
+
+        relations = interfaces[relation_name]
+        if not isinstance(relations, SerializedDataInterface):
+            raise CheckFailed(f"Unexpected error with {relation_name} relation data - data not as expected", BlockedStatus)
```

Comment:
This again seems like something that I prefer erroring out instead of us handling it.

Reply:
This should be resolved the same as the above conversation.  In this case, I am not sure this can happen if SDI works correctly anyway.